File: springboot-crud/src/main/java/com/example/crud/service/CategoryService.java
package com.example.crud.service;

import com.example.crud.dto.CategoryDTO;
import com.example.crud.entity.Category;
import com.example.crud.exception.ResourceNotFoundException;
import com.example.crud.repository.CategoryRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class CategoryService {

    private final CategoryRepository categoryRepository;

    public CategoryService(CategoryRepository categoryRepository) {
        this.categoryRepository = categoryRepository;
    }

    // Create
    public CategoryDTO createCategory(CategoryDTO dto) {
        Category category = new Category();
        category.setName(dto.getName());
        Category saved = categoryRepository.save(category);
        return new CategoryDTO(saved.getId(), saved.getName());
    }

    // Read all
    public List<CategoryDTO> getAllCategories() {
        return categoryRepository.findAll().stream()
                .map(c -> new CategoryDTO(c.getId(), c.getName()))
                .collect(Collectors.toList());
    }

    // Read by ID
    public CategoryDTO getCategoryById(Long id) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found with id: " + id));
        return new CategoryDTO(category.getId(), category.getName());
    }

    // Update
    public CategoryDTO updateCategory(Long id, CategoryDTO dto) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found with id: " + id));
        category.setName(dto.getName());
        Category updated = categoryRepository.save(category);
        return new CategoryDTO(updated.getId(), updated.getName());
    }

    // Delete
    public void deleteCategory(Long id) {
        if (!categoryRepository.existsById(id)) {
            throw new ResourceNotFoundException("Category not found with id: " + id);
        }
        categoryRepository.deleteById(id);
    }
}

File: frontend/src/services/categoryService.js
import api, { API_BASE_URL } from "./api";

// Lấy tất cả categories
export const getCategories = async () => {
    const res = await api.get("/categories");
    return res.data;
};

// Tạo mới category
export const createCategory = async (category) => {
    const res = await api.post("/categories", category);
    return res.data;
};

// Cập nhật category
export const updateCategory = async (id, category) => {
    const res = await api.put(`/categories/${id}`, category);
    return res.data;
};

// Xóa category
export const deleteCategory = async (id) => {
    const res = await api.delete(`/categories/${id}`);
    return res.data;
};
